lightning/src/ln/channel.rs

```diff
@@ -4217,6 +4208,48 @@ impl<Signer: Sign> Channel<Signer> {
 	}
 }
 
+fn is_unsupported_shutdown_script(their_features: &InitFeatures, script: &Script) -> bool {
+	// We restrain shutdown scripts to standards forms to avoid transactions not propagating on the p2p tx-relay network
+
+	// Providing an empty is supported
+	let script_lenght = script.len();
+	if script_lenght == 0 {
+		return false;
+	}
+
+	// BOLT 2 says we must only send a scriptpubkey of certain standard forms,
+	// which for a a BIP-141-compliant witness program is at max 42 bytes in length.
+	// So don't let the remote peer feed us some super fee-heavy script.
+	let is_script_too_long = script_lenght > 42;
+	if is_script_too_long {
+		return true;
+	}
+
+	if their_features.supports_shutdown_anysegwit() {
+		// A scriptPubKey (or redeemScript as defined in BIP16/P2SH) that consists of a 1-byte
+		// push opcode (for 1 to 16) followed by a data push between 2 and 40 bytes gets a new
+		// special meaning.
+		let min_vernum: u8 = opcodes::all::OP_PUSHNUM_1.into_u8();
+		let max_vernum: u8 = opcodes::all::OP_PUSHNUM_16.into_u8();
+		let script_bytes = script.as_bytes();
+		let is_anysegwit = script.len() >= 3
+			&& script.len() <= 42
+			// PUSHNUM_1-PUSHNUM_16
+			&& (script_bytes[0] >= min_vernum && script_bytes[0] <= max_vernum)
+			// Second byte push opcode 2-40 bytes
+			&& script_bytes[1] >= opcodes::all::OP_PUSHBYTES_2.into_u8()
```

Comment:
Why not allow PUSHBYTES_1? More generally, instead of adding a bunch of script parsing code, can we not just switch to `script.is_witness_script() && script.as_bytes()[0] != OP_PUSHNUM_1`? That should be less error-prone.